File: 103.py
#!/usr/bin/env python

"""This script solves the Project Euler problem "Special subset sums: optimum".
The problem is: Given that A is an optimum special sum set for n = 7, find its
set string.
"""

from __future__ import division
import argparse
from itertools import product, combinations, chain


def main(args):
    """Special subset sums: optimum"""

    full_set = set([1])
    for n in range(2, args.set_size + 1):
        full_set = algorithm_set(full_set)
        full_set = optimum_set(full_set)

    print(''.join(map(str, sorted(full_set))))


def algorithm_set(prev_set):
    """Get set from previous set using algorithm"""

    middle = sorted(prev_set)[len(prev_set) // 2]
    new_set = set([middle])
    for i in sorted(prev_set):
        new_set.add(middle + i)

    return new_set


def optimum_set(algo_set):
    """Get optimum set from algorithm set using nearby search"""

    n = len(algo_set)

    opt_sum = sum(algo_set)
    opt_set = algo_set
    for offset in product(range(-2, 3), repeat=n):
        test_set = set([i + j for i, j in zip(sorted(algo_set), offset)])

        if sum(test_set) > opt_sum:
            continue
        if len(test_set) != n:
            continue
        if any(i <= 0 for i in test_set):
            continue

        subsets = list(chain.from_iterable(combinations(test_set, i)
                                           for i in range(n + 1)))
        subsets = [set(sub) for sub in subsets]
        subsets = [sub for sub in subsets if len(sub) > 0 and len(sub) < n]

        sums = [sum(sub) for sub in subsets]
        lens = [len(sub) for sub in subsets]

        # Check for disjoint subsets with equal sums
        got_disjoint_eq_sum = False
        for subset_sum in set(sums):
            indices = [i for i, s in enumerate(sums) if s == subset_sum]
            if len(indices) == 1:
                continue
            for i, j in combinations(indices, 2):
                if not len(subsets[i].intersection(subsets[j])):
                    got_disjoint_eq_sum = True
                    break
        if got_disjoint_eq_sum:
            continue

        # Check for disjoint subsets of different lengths and sums
        got_disjoint_diff = False
        for i, j in combinations(range(len(subsets)), 2):
            if lens[i] == lens[j] or sums[i] == sums[j]:
                continue
            if len(subsets[i].intersection(subsets[j])):
                continue
            if lens[i] > lens[j] and sums[i] < sums[j]:
                got_disjoint_diff = True
                break
            if lens[j] > lens[i] and sums[j] < sums[i]:
                got_disjoint_diff = True
                break
        if got_disjoint_diff:
            continue

        opt_sum = sum(test_set)
        opt_set = test_set

    return opt_set

if __name__ == '__main__':
    parser = argparse.ArgumentParser(
        description='Special subset sums: optimum')
    parser.add_argument(
        'set_size', metavar='SET_SIZE', type=int, default=7, nargs='?',
        help='The set size')
    args = parser.parse_args()

    main(args)

File: 103.pl
#!/usr/bin/env perl

# PODNAME: 103.pl
# ABSTRACT: Special subset sums: optimum

## Author     : ian.sealy@sanger.ac.uk
## Maintainer : ian.sealy@sanger.ac.uk
## Created    : 2016-05-24

use warnings;
use strict;
use autodie;
use Getopt::Long;
use Pod::Usage;
use Carp;
use version; our $VERSION = qv('v0.1.0');

use Readonly;
use List::Util qw(sum notall any);
use Algorithm::Combinatorics qw(variations_with_repetition combinations);

# Constants
Readonly our $OFFSET => 2;

# Default options
my $set_size = 7;    ## no critic (ProhibitMagicNumbers)
my ( $debug, $help, $man );

# Get and check command line options
get_and_check_options();

my @set = (1);       ## no critic (ProhibitAmbiguousNames)
foreach my $n ( 2 .. $set_size ) {
    @set = algorithm_set(@set);
    @set = optimum_set(@set);
}

printf "%s\n", join q{}, @set;

# Get set from previous set using algorithm
sub algorithm_set {
    my (@previous_set) = @_;

    my $middle = @previous_set[ int( scalar @previous_set ) / 2 ];
    ## no critic (ProhibitAmbiguousNames, ProhibitReusedNames)
    my @set = ($middle);
    ## use critic
    foreach my $int (@previous_set) {
        push @set, $middle + $int;
    }

    return @set;
}

# Get optimum set from algorithm set using nearby search
sub optimum_set {
    my (@algo_set) = @_;

    my $n = scalar @algo_set;

    # Make indices for subsets
    my @subset_idxs = variations_with_repetition( [ 0, 1 ], $n );
    @subset_idxs = grep {
        notall { $_ == 1 }
        @{$_}
    } @subset_idxs;
    @subset_idxs = grep {
        notall { $_ == 0 }
        @{$_}
    } @subset_idxs;

    # Check nearby sets
    my $optimum_sum = sum(@algo_set);
    my @optimum_set = @algo_set;
    my $offset_iterator =
      variations_with_repetition( [ -$OFFSET .. $OFFSET ], $n );
  SET: while ( my $offset = $offset_iterator->next ) {
        ## no critic (ProhibitAmbiguousNames, ProhibitReusedNames)
        my @set = map { $algo_set[$_] + $offset->[$_] } ( 0 .. $n - 1 );
        ## use critic
        next SET if any { $_ <= 0 } @set;    # Ensure positive integers

        # Ensure unique integers
        my %seen;
        foreach my $int (@set) {
            next SET if exists $seen{$int};
            $seen{$int} = 1;
        }

        # Skip definitely non-optimal sets
        my $set_sum = sum(@set);    ## no critic (ProhibitAmbiguousNames)
        next SET if $set_sum > $optimum_sum;

        # Get all subsets
        my %subsets_by_sum;
        my @subsets;
        my @sums;
        my @lengths;
        foreach my $subset_idx (@subset_idxs) {
            ## no critic (ProhibitAmbiguousNames)
            my @subset =
              @set[ grep { $subset_idx->[$_] == 1 } ( 0 .. $n - 1 ) ];
            ## use critic
            my $subset_sum = sum(@subset);
            push @{ $subsets_by_sum{$subset_sum} }, \@subset;
            push @subsets, \@subset;
            push @sums,    $subset_sum;
            push @lengths, scalar @subset;
        }

        # Check for disjoint subsets with equal sums
        foreach my $sum ( keys %subsets_by_sum ) {
            next if scalar @{ $subsets_by_sum{$sum} } == 1;
            foreach my $pair ( combinations( $subsets_by_sum{$sum}, 2 ) ) {
                next SET if disjoint( $pair->[0], $pair->[1] );
            }
        }

        # Check for disjoint subsets of different lengths and sums
        my $pair_iterator = combinations( [ 0 .. scalar @subsets - 1 ], 2 );
        while ( my $pair = $pair_iterator->next ) {
            my ( $i, $j ) = @{$pair};
            next if $lengths[$i] == $lengths[$j];
            next if $sums[$i] == $sums[$j];
            next SET
              if ( ( $lengths[$i] > $lengths[$j] && $sums[$i] < $sums[$j] )
                || ( $lengths[$j] > $lengths[$i] && $sums[$j] < $sums[$i] ) )
              && disjoint( $subsets[$i], $subsets[$j] );
        }

        $optimum_sum = $set_sum;
        @optimum_set = sort { $a <=> $b } @set;
    }

    return @optimum_set;
}

sub disjoint {
    my ( $set1, $set2 ) = @_;

    my %seen;
    map { $seen{$_}++ } @{$set1};
    map { $seen{$_}++ } @{$set2};

    if ( notall { $_ == 1 } values %seen ) {
        return 0;
    }
    return 1;
}

# Get and check command line options
sub get_and_check_options {

    # Get options
    GetOptions(
        'set_size=i' => \$set_size,
        'debug'      => \$debug,
        'help'       => \$help,
        'man'        => \$man,
    ) or pod2usage(2);

    # Documentation
    if ($help) {
        pod2usage(1);
    }
    elsif ($man) {
        pod2usage( -verbose => 2 );
    }

    return;
}

__END__
=pod

=encoding UTF-8

=head1 NAME

103.pl

Special subset sums: optimum

=head1 VERSION

version 0.1.0

=head1 DESCRIPTION

This script solves the Project Euler problem "Special subset sums: optimum". The
problem is: Given that A is an optimum special sum set for n = 7, find its set
string.

=head1 EXAMPLES

    perl 103.pl --set_size 5

=head1 USAGE

    103.pl
        [--set_size INT]
        [--debug]
        [--help]
        [--man]

=head1 OPTIONS

=over 8

=item B<--set_size INT>

The set size.

=item B<--debug>

Print debugging information.

=item B<--help>

Print a brief help message and exit.

=item B<--man>

Print this script's manual page and exit.

=back

=head1 DEPENDENCIES

None

=head1 AUTHOR

=over 4

=item *

Ian Sealy <ian.sealy@sanger.ac.uk>

=back

=head1 COPYRIGHT AND LICENSE

This software is Copyright (c) 2016 by Genome Research Ltd.

This is free software, licensed under:

  The GNU General Public License, Version 3, June 2007

=cut
